File: packages/python/src/unmsm_sum_api/http.py
from typing import Any, Callable, TypeVar

import httpx
from pydantic import BaseModel

from .errors import SumApiError

TModel = TypeVar("TModel", bound=BaseModel)


class HttpClient:
    def __init__(self, *, base_url: str, client: httpx.Client, get_token: Callable[[], str | None]) -> None:
        self.base_url = base_url.rstrip("/")
        self.client = client
        self.get_token = get_token

    def request(
        self,
        path: str,
        model: type[TModel],
        *,
        method: str = "GET",
        json: dict[str, Any] | None = None,
        auth: bool = True,
    ) -> TModel:
        headers = {"accept": "application/json"}
        if auth:
            token = self.get_token()
            if not token:
                raise RuntimeError("Authentication token is required. Call login() or pass token in the constructor.")
            headers["authorization"] = f"Bearer {token}"

        response = self.client.request(method, f"{self.base_url}{path}", headers=headers, json=json)
        body = response.json() if response.content else None

        if response.is_error:
            raise SumApiError(f"SUM API request failed with status {response.status_code}", response.status_code, body)

        return model.model_validate(body)

File: packages/js/src/http.ts
import type { z } from "zod";

import { SumApiError } from "./errors.js";

export type HttpClientOptions = {
  baseUrl: string;
  fetch: typeof fetch;
  getToken: () => string | undefined;
};

export type RequestOptions = {
  method?: string;
  body?: unknown;
  auth?: boolean;
};

export class HttpClient {
  constructor(private readonly options: HttpClientOptions) {}

  async request<TSchema extends z.ZodTypeAny>(
    path: string,
    schema: TSchema,
    options: RequestOptions = {},
  ): Promise<z.infer<TSchema>> {
    const headers: Record<string, string> = {
      accept: "application/json",
    };

    if (options.body !== undefined) headers["content-type"] = "application/json";
    if (options.auth !== false) {
      const token = this.options.getToken();
      if (!token) throw new Error("Authentication token is required. Call login() or pass token in the constructor.");
      headers.authorization = `Bearer ${token}`;
    }

    const response = await this.options.fetch(`${this.options.baseUrl}${path}`, {
      method: options.method ?? "GET",
      headers,
      body: options.body === undefined ? undefined : JSON.stringify(options.body),
    });

    const text = await response.text();
    const body = text ? JSON.parse(text) : null;

    if (!response.ok) {
      throw new SumApiError(`SUM API request failed with status ${response.status}`, response.status, body);
    }

    return schema.parse(body);
  }
}
